app/controllers/incident/AddEndorsementController.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.incident
+
+import controllers.actions._
+import controllers.{NavigatorOps, SettableOps, SettableOpsRunner}
+import forms.YesNoFormProvider
+import models.{Index, Mode, MovementReferenceNumber}
+import navigation.{IncidentNavigator, IncidentNavigatorProvider}
+import pages.incident.AddEndorsementPage
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import repositories.SessionRepository
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import views.html.incident.AddEndorsementView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class AddEndorsementController @Inject() (
+  override val messagesApi: MessagesApi,
+  implicit val sessionRepository: SessionRepository,
+  navigatorProvider: IncidentNavigatorProvider,
+  actions: Actions,
+  formProvider: YesNoFormProvider,
+  val controllerComponents: MessagesControllerComponents,
+  view: AddEndorsementView
+)(implicit ec: ExecutionContext)
+    extends FrontendBaseController
+    with I18nSupport {
+
+  private val form = formProvider("Incident.addEndorsement")
+
+  def onPageLoad(mrn: MovementReferenceNumber, mode: Mode, index: Index): Action[AnyContent] = actions.requireData(mrn) {
+    implicit request =>
+      val preparedForm = request.userAnswers.get(AddEndorsementPage(index)) match {
+        case None        => form
+        case Some(value) => form.fill(value)
+      }
+
+      Ok(view(preparedForm, mrn, mode, index))
+  }
+
+  def onSubmit(mrn: MovementReferenceNumber, mode: Mode, index: Index): Action[AnyContent] = actions.requireData(mrn).async {
+    implicit request =>
+      form
+        .bindFromRequest()
+        .fold(
+          formWithErrors => Future.successful(BadRequest(view(formWithErrors, mrn, mode, index))),
+          value => {
+            implicit val navigator: IncidentNavigator = navigatorProvider(index)
```

Comment:
Hi Andy. I do like this idea as im not a big fan of making implicit vals in scope. We could manually add them into the implicit however if all the index's are the same maybe we are better off doing something like this in the navigator ops package:

`    def navigateWith(mode: Mode)(implicit navigator: Navigator, executionContext: ExecutionContext): Future[Result] =
      navigate {
        case (_, userAnswers) => navigator.nextPage(userAnswers, mode)
      }`

Reply:
In a few cases the navigator can be injected as an implicit val into the controller. It's only when some additional parameters are needed (an index, or some reference data lists, or maybe both) that we need to declare the val in scope. I don't mind it. As vals go I think it's a good candidate for an implicit as we are only ever going to have one instance of it per controller (same with the SessionRepository). It was more a stylistic choice than anything to avoid having to pass it in explicitly everywhere to make the code a bit more readable.

Reply:
Do the suggested changes need to come in with this PR?

Reply:
I think given @jackheslop96 explanation we can leave as it is. Im not a massive fan of implicits in this context however, I do understand the argument here. :-) . Will approve.